lib/gc/Transforms/GPU/GpuTilingAndFusion.cpp

```diff
@@ -0,0 +1,481 @@
+              }
+            }
+          }
+
+          assert(!sizes.empty());
+          auto elementSize = getElementSize(op);
+          auto sizePerThread = numIterations / numThreads * elementSize;
+          auto totalSize = std::max(sizePerThread, cachePerThread);
+          totalSize = std::max(totalSize / elementSize, 64L);
+          int64_t minTileSize = 1;
+
+          // If the operation could be lowered to XeGPU, make the tiles
+          // multiple of the vector width and the minimum tile size 8.
+          if (canLowerToXeGPU(op)) {
+            minTileSize = 8;
+            totalSize = std::max(totalSize / vectorWidth, 1L) * vectorWidth;
+          }
+
+          SmallVector<int64_t> tiles = sizes;
+          adjustTiles(totalSize, tiles, minTileSize);
+
+          // If the tiles are not less than the sizes, split the largest tile
+          // into two to avoid loops elimination by the canonicalizer pass.
+          if (auto pairs = zip(sizes, tiles);
+              std::all_of(pairs.begin(), pairs.end(), [](auto p) {
+                return std::get<0>(p) <= std::get<1>(p);
+              })) {
+            for (auto &t : tiles) {
+              if (t >= maxSize) {
+                t = std::max(1L, floorPow2(maxSize / 2));
+                break;
+              }
+            }
+          }
+
+          unsigned counter = 0;
+          SmallVector<OpFoldResult> result;
+          result.reserve(itDomains.size());
+
+          for (auto [t, r] : zip(itTypes, itDomains)) {
+            if (t == utils::IteratorType::parallel) {
+              result.emplace_back(rw.getConstant(tiles[counter++]));
+            } else {
+              result.emplace_back(rw.getConstant(0));
+            }
+          }
+
+          return result;
+        });
+    opts.setFusionControlFn(
+        [&](tensor::ExtractSliceOp candidateSliceOp, OpResult originalProducer,
+            bool) -> std::optional<SCFTileAndFuseOptions::ControlFnResult> {
+          Operation *op = originalProducer.getOwner();
+          if (!op) {
+            return std::nullopt;
+          }
+          if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+            if (!linalgOp.hasOnlyProjectedPermutations()) {
+              return std::nullopt;
+            }
+          }
+
+          // If the result of this slice is used by a MatmulOp and the slice has
+          // an operand produced by a previous MatmulOp, do not fuse.
+          if (isOpDependsOnResult<0>(linalgx::isMatmulOp, candidateSliceOp) &&
+              isOperandDependsOnOp(linalgx::isMatmulOp, candidateSliceOp)) {
+            return std::nullopt;
+          }
+
+          return SCFTileAndFuseOptions::ControlFnResult{};
+        });
+    opts.tilingOptions.setLoopType(SCFTilingOptions::LoopType::ForallOp);
+
+    for (auto ti = findTi(rw, fn); ti; ti = findTi(rw, fn)) {
+      auto result = tileConsumerAndFuseProducersUsingSCF(rw, *ti, opts);
```

Comment:
Why use `tileConsumerAndFuseProducersUsingSCF` instead of `tileProducerAndFuseConsumerUsingSCF` here and how to choose the tiling target op here? For the fusion `matmul + add + relu`, we should tile the relu as the `add` and `matmul` is relu's producer.

Reply:
The `fuse consumer` and `fuse producer` are all upstream avaliable function(https://mlir.llvm.org/docs/Dialects/Transform/#transformstructuredfuse_into_containing_op-transformfuseintocontainingop, https://mlir.llvm.org/docs/Dialects/Transform/#transformstructuredfuse-transformfuseop). The point is why choose one instead of another and how to choose the target op. If you call `fuse producer`(`fuse consumer`), you need to find out the last op(first op) in the fusion chain.

Reply:
I choose tileConsumerAndFuseProducersUsingSCF instead of tileProducerAndFuseConsumerUsingSCF because there is no such function. The operations are processed in the bottom-up order, so we start from the last op.

Reply:
The name is not exactly the same. The function name should be `tileAndFuseConsumerOfSlice` in https://github.com/llvm/llvm-project/pull/88712. Yes, we should start from the last op but how do you decide which op is the last op. For example, there will be a branch in the graph and how do you determine what op is the last op(`add` or `relu`) and this is part of the reason why the `tile consumer` is proposed.
```
        ->add  
      /
matmul -> relu 

```

Reply:
In this case matmul could be fused into each branch. Look at the following examples and compare the results.  GpuTilingAndFusion moves all the operations into the branches, but IterativeTilingAndFusion does not fuse at all.

<details>
<summary>Example 1</summary>

```MLIR
  func.func @test1(%arg0: memref<256x256xf16>, %arg1: memref<256x256xf16>, %arg2: memref<256x256xf16>, %cond: i1) {
    %t0 = bufferization.to_tensor %arg0 restrict : memref<256x256xf16>
    %t1 = bufferization.to_tensor %arg1 restrict : memref<256x256xf16>
    %tmp1 = tensor.empty() : tensor<256x256xf16>
    %0 = linalg.matmul ins(%t0, %t1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp1 : tensor<256x256xf16>) -> tensor<256x256xf16>
    %1 = scf.if %cond -> tensor<256x256xf16> {
      %tmp2 = tensor.empty() : tensor<256x256xf16>
      %2 = linalg.add ins(%0, %t1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp2 : tensor<256x256xf16>) -> tensor<256x256xf16>
      scf.yield %2 : tensor<256x256xf16>
    } else {
      %tmp3 = tensor.empty() : tensor<256x256xf16>
      %3 = linalg.mul ins(%0, %t1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp3 : tensor<256x256xf16>) -> tensor<256x256xf16>
      scf.yield %3 : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %1 in restrict writable %arg2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>
<details>
<summary>GpuTilingAndFusion</summary>

```MLIR
  func.func @test1(%arg0: memref<256x256xf16>, %arg1: memref<256x256xf16>, %arg2: memref<256x256xf16>, %arg3: i1, %arg4: memref<i8>) {
    %0 = bufferization.to_tensor %arg0 restrict : memref<256x256xf16>
    %1 = bufferization.to_tensor %arg1 restrict : memref<256x256xf16>
    %2 = scf.if %arg3 -> (tensor<256x256xf16>) {
      %3 = tensor.empty() : tensor<256x256xf16>
      %4 = scf.forall (%arg5) = (0) to (256) step (128) shared_outs(%arg6 = %3) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<256x256xf16> to tensor<128x256xf16>
        %5 = scf.forall (%arg7, %arg8) = (0, 0) to (128, 256) step (16, 32) shared_outs(%arg9 = %extracted_slice) -> (tensor<128x256xf16>) {
          %6 = arith.addi %arg7, %arg5 : index
          %extracted_slice_0 = tensor.extract_slice %0[%6, 0] [16, 256] [1, 1] : tensor<256x256xf16> to tensor<16x256xf16>
          %extracted_slice_1 = tensor.extract_slice %1[0, %arg8] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
          %7 = tensor.empty() : tensor<16x32xf16>
          %8 = linalg.matmul ins(%extracted_slice_0, %extracted_slice_1 : tensor<16x256xf16>, tensor<256x32xf16>) outs(%7 : tensor<16x32xf16>) -> tensor<16x32xf16>
          %extracted_slice_2 = tensor.extract_slice %1[%6, %arg8] [16, 32] [1, 1] : tensor<256x256xf16> to tensor<16x32xf16>
          %extracted_slice_3 = tensor.extract_slice %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<128x256xf16> to tensor<16x32xf16>
          %9 = linalg.add ins(%8, %extracted_slice_2 : tensor<16x32xf16>, tensor<16x32xf16>) outs(%extracted_slice_3 : tensor<16x32xf16>) -> tensor<16x32xf16>
          scf.forall.in_parallel {
            tensor.parallel_insert_slice %9 into %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<16x32xf16> into tensor<128x256xf16>
          }
        }
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %5 into %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<128x256xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %4 : tensor<256x256xf16>
    } else {
      %3 = tensor.empty() : tensor<256x256xf16>
      %4 = scf.forall (%arg5) = (0) to (256) step (128) shared_outs(%arg6 = %3) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<256x256xf16> to tensor<128x256xf16>
        %5 = scf.forall (%arg7, %arg8) = (0, 0) to (128, 256) step (16, 32) shared_outs(%arg9 = %extracted_slice) -> (tensor<128x256xf16>) {
          %6 = arith.addi %arg7, %arg5 : index
          %extracted_slice_0 = tensor.extract_slice %0[%6, 0] [16, 256] [1, 1] : tensor<256x256xf16> to tensor<16x256xf16>
          %extracted_slice_1 = tensor.extract_slice %1[0, %arg8] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
          %7 = tensor.empty() : tensor<16x32xf16>
          %8 = linalg.matmul ins(%extracted_slice_0, %extracted_slice_1 : tensor<16x256xf16>, tensor<256x32xf16>) outs(%7 : tensor<16x32xf16>) -> tensor<16x32xf16>
          %extracted_slice_2 = tensor.extract_slice %1[%6, %arg8] [16, 32] [1, 1] : tensor<256x256xf16> to tensor<16x32xf16>
          %extracted_slice_3 = tensor.extract_slice %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<128x256xf16> to tensor<16x32xf16>
          %9 = linalg.mul ins(%8, %extracted_slice_2 : tensor<16x32xf16>, tensor<16x32xf16>) outs(%extracted_slice_3 : tensor<16x32xf16>) -> tensor<16x32xf16>
          scf.forall.in_parallel {
            tensor.parallel_insert_slice %9 into %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<16x32xf16> into tensor<128x256xf16>
          }
        }
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %5 into %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<128x256xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %4 : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %2 in restrict writable %arg2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>
<details>
<summary>IterativeTilingAndFusion</summary>

```MLIR
  func.func @test1(%arg0: memref<256x256xf16>, %arg1: memref<256x256xf16>, %arg2: memref<256x256xf16>, %arg3: i1, %arg4: memref<i8>) {
    %0 = bufferization.to_tensor %arg0 restrict : memref<256x256xf16>
    %1 = bufferization.to_tensor %arg1 restrict : memref<256x256xf16>
    %2 = tensor.empty() : tensor<256x256xf16>
    %3 = scf.forall (%arg5, %arg6) = (0, 0) to (256, 256) step (32, 32) shared_outs(%arg7 = %2) -> (tensor<256x256xf16>) {
      %extracted_slice = tensor.extract_slice %0[%arg5, 0] [32, 256] [1, 1] : tensor<256x256xf16> to tensor<32x256xf16>
      %extracted_slice_0 = tensor.extract_slice %1[0, %arg6] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
      %extracted_slice_1 = tensor.extract_slice %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
      %5 = linalg.matmul ins(%extracted_slice, %extracted_slice_0 : tensor<32x256xf16>, tensor<256x32xf16>) outs(%extracted_slice_1 : tensor<32x32xf16>) -> tensor<32x32xf16>
      scf.forall.in_parallel {
        tensor.parallel_insert_slice %5 into %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<32x32xf16> into tensor<256x256xf16>
      }
    }
    %4 = scf.if %arg3 -> (tensor<256x256xf16>) {
      %5 = tensor.empty() : tensor<256x256xf16>
      %6 = scf.forall (%arg5, %arg6) = (0, 0) to (256, 256) step (32, 32) shared_outs(%arg7 = %5) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %3[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_0 = tensor.extract_slice %1[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_1 = tensor.extract_slice %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %7 = linalg.add ins(%extracted_slice, %extracted_slice_0 : tensor<32x32xf16>, tensor<32x32xf16>) outs(%extracted_slice_1 : tensor<32x32xf16>) -> tensor<32x32xf16>
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %7 into %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<32x32xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %6 : tensor<256x256xf16>
    } else {
      %5 = tensor.empty() : tensor<256x256xf16>
      %6 = scf.forall (%arg5, %arg6) = (0, 0) to (256, 256) step (32, 32) shared_outs(%arg7 = %5) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %3[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_0 = tensor.extract_slice %1[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_1 = tensor.extract_slice %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %7 = linalg.mul ins(%extracted_slice, %extracted_slice_0 : tensor<32x32xf16>, tensor<32x32xf16>) outs(%extracted_slice_1 : tensor<32x32xf16>) -> tensor<32x32xf16>
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %7 into %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<32x32xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %6 : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %4 in restrict writable %arg2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>
<details>
<summary>Example 2</summary>

```MLIR
  func.func @test2(%memref0: memref<256x256xf16>, %memref1: memref<256x256xf16>, %memref2: memref<256x256xf16>, %cond: i1) {
    %arg0 = bufferization.to_tensor %memref0 restrict : memref<256x256xf16>
    %arg1 = bufferization.to_tensor %memref1 restrict : memref<256x256xf16>
    %tmp1 = tensor.empty() : tensor<256x256xf16>
    %matmul_res = linalg.matmul ins(%arg0, %arg1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp1 : tensor<256x256xf16>) -> tensor<256x256xf16>
    %tmp2 = tensor.empty() : tensor<256x256xf16>
    %add_res = linalg.add ins(%matmul_res, %arg1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp2 : tensor<256x256xf16>) -> tensor<256x256xf16>
    %if_res = scf.if %cond -> tensor<256x256xf16> {
      scf.yield %add_res : tensor<256x256xf16>
    } else {
      %tmp3 = tensor.empty() : tensor<256x256xf16>
      %mul_res = linalg.mul ins(%add_res, %arg1 : tensor<256x256xf16>, tensor<256x256xf16>) outs(%tmp3 : tensor<256x256xf16>) -> tensor<256x256xf16>
      scf.yield %mul_res : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %if_res in restrict writable %memref2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>
<details>
<summary>GpuTilingAndFusion</summary>

```MLIR
  func.func @test2(%arg0: memref<256x256xf16>, %arg1: memref<256x256xf16>, %arg2: memref<256x256xf16>, %arg3: i1, %arg4: memref<i8>) {
    %0 = bufferization.to_tensor %arg0 restrict : memref<256x256xf16>
    %1 = bufferization.to_tensor %arg1 restrict : memref<256x256xf16>
    %2 = tensor.empty() : tensor<256x256xf16>
    %3 = scf.if %arg3 -> (tensor<256x256xf16>) {
      %4 = scf.forall (%arg5) = (0) to (256) step (128) shared_outs(%arg6 = %2) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<256x256xf16> to tensor<128x256xf16>
        %5 = scf.forall (%arg7, %arg8) = (0, 0) to (128, 256) step (16, 32) shared_outs(%arg9 = %extracted_slice) -> (tensor<128x256xf16>) {
          %6 = arith.addi %arg7, %arg5 : index
          %extracted_slice_0 = tensor.extract_slice %0[%6, 0] [16, 256] [1, 1] : tensor<256x256xf16> to tensor<16x256xf16>
          %extracted_slice_1 = tensor.extract_slice %1[0, %arg8] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
          %7 = tensor.empty() : tensor<16x32xf16>
          %8 = linalg.matmul ins(%extracted_slice_0, %extracted_slice_1 : tensor<16x256xf16>, tensor<256x32xf16>) outs(%7 : tensor<16x32xf16>) -> tensor<16x32xf16>
          %extracted_slice_2 = tensor.extract_slice %1[%6, %arg8] [16, 32] [1, 1] : tensor<256x256xf16> to tensor<16x32xf16>
          %extracted_slice_3 = tensor.extract_slice %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<128x256xf16> to tensor<16x32xf16>
          %9 = linalg.add ins(%8, %extracted_slice_2 : tensor<16x32xf16>, tensor<16x32xf16>) outs(%extracted_slice_3 : tensor<16x32xf16>) -> tensor<16x32xf16>
          scf.forall.in_parallel {
            tensor.parallel_insert_slice %9 into %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<16x32xf16> into tensor<128x256xf16>
          }
        }
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %5 into %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<128x256xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %4 : tensor<256x256xf16>
    } else {
      %4 = tensor.empty() : tensor<256x256xf16>
      %5 = scf.forall (%arg5) = (0) to (256) step (128) shared_outs(%arg6 = %4) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<256x256xf16> to tensor<128x256xf16>
        %6 = scf.forall (%arg7, %arg8) = (0, 0) to (128, 256) step (16, 32) shared_outs(%arg9 = %extracted_slice) -> (tensor<128x256xf16>) {
          %7 = arith.addi %arg7, %arg5 : index
          %extracted_slice_0 = tensor.extract_slice %0[%7, 0] [16, 256] [1, 1] : tensor<256x256xf16> to tensor<16x256xf16>
          %extracted_slice_1 = tensor.extract_slice %1[0, %arg8] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
          %8 = tensor.empty() : tensor<16x32xf16>
          %9 = linalg.matmul ins(%extracted_slice_0, %extracted_slice_1 : tensor<16x256xf16>, tensor<256x32xf16>) outs(%8 : tensor<16x32xf16>) -> tensor<16x32xf16>
          %extracted_slice_2 = tensor.extract_slice %1[%7, %arg8] [16, 32] [1, 1] : tensor<256x256xf16> to tensor<16x32xf16>
          %10 = tensor.empty() : tensor<16x32xf16>
          %11 = linalg.add ins(%9, %extracted_slice_2 : tensor<16x32xf16>, tensor<16x32xf16>) outs(%10 : tensor<16x32xf16>) -> tensor<16x32xf16>
          %extracted_slice_3 = tensor.extract_slice %1[%7, %arg8] [16, 32] [1, 1] : tensor<256x256xf16> to tensor<16x32xf16>
          %extracted_slice_4 = tensor.extract_slice %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<128x256xf16> to tensor<16x32xf16>
          %12 = linalg.mul ins(%11, %extracted_slice_3 : tensor<16x32xf16>, tensor<16x32xf16>) outs(%extracted_slice_4 : tensor<16x32xf16>) -> tensor<16x32xf16>
          scf.forall.in_parallel {
            tensor.parallel_insert_slice %12 into %arg9[%arg7, %arg8] [16, 32] [1, 1] : tensor<16x32xf16> into tensor<128x256xf16>
          }
        }
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %6 into %arg6[%arg5, 0] [128, 256] [1, 1] : tensor<128x256xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %5 : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %3 in restrict writable %arg2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>
<details>
<summary>IterativeTilingAndFusion</summary>

```MLIR
  func.func @test2(%arg0: memref<256x256xf16>, %arg1: memref<256x256xf16>, %arg2: memref<256x256xf16>, %arg3: i1, %arg4: memref<i8>) {
    %0 = bufferization.to_tensor %arg0 restrict : memref<256x256xf16>
    %1 = bufferization.to_tensor %arg1 restrict : memref<256x256xf16>
    %2 = tensor.empty() : tensor<256x256xf16>
    %3 = tensor.empty() : tensor<256x256xf16>
    %4 = scf.forall (%arg5, %arg6) = (0, 0) to (256, 256) step (32, 32) shared_outs(%arg7 = %3) -> (tensor<256x256xf16>) {
      %extracted_slice = tensor.extract_slice %0[%arg5, 0] [32, 256] [1, 1] : tensor<256x256xf16> to tensor<32x256xf16>
      %extracted_slice_0 = tensor.extract_slice %1[0, %arg6] [256, 32] [1, 1] : tensor<256x256xf16> to tensor<256x32xf16>
      %extracted_slice_1 = tensor.extract_slice %2[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
      %6 = linalg.matmul ins(%extracted_slice, %extracted_slice_0 : tensor<32x256xf16>, tensor<256x32xf16>) outs(%extracted_slice_1 : tensor<32x32xf16>) -> tensor<32x32xf16>
      %extracted_slice_2 = tensor.extract_slice %1[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
      %extracted_slice_3 = tensor.extract_slice %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
      %7 = linalg.add ins(%6, %extracted_slice_2 : tensor<32x32xf16>, tensor<32x32xf16>) outs(%extracted_slice_3 : tensor<32x32xf16>) -> tensor<32x32xf16>
      scf.forall.in_parallel {
        tensor.parallel_insert_slice %7 into %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<32x32xf16> into tensor<256x256xf16>
      }
    }
    %5 = scf.if %arg3 -> (tensor<256x256xf16>) {
      scf.yield %4 : tensor<256x256xf16>
    } else {
      %6 = tensor.empty() : tensor<256x256xf16>
      %7 = scf.forall (%arg5, %arg6) = (0, 0) to (256, 256) step (32, 32) shared_outs(%arg7 = %6) -> (tensor<256x256xf16>) {
        %extracted_slice = tensor.extract_slice %4[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_0 = tensor.extract_slice %1[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %extracted_slice_1 = tensor.extract_slice %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<256x256xf16> to tensor<32x32xf16>
        %8 = linalg.mul ins(%extracted_slice, %extracted_slice_0 : tensor<32x32xf16>, tensor<32x32xf16>) outs(%extracted_slice_1 : tensor<32x32xf16>) -> tensor<32x32xf16>
        scf.forall.in_parallel {
          tensor.parallel_insert_slice %8 into %arg7[%arg5, %arg6] [32, 32] [1, 1] : tensor<32x32xf16> into tensor<256x256xf16>
        }
      }
      scf.yield %7 : tensor<256x256xf16>
    }
    bufferization.materialize_in_destination %5 in restrict writable %arg2 : (tensor<256x256xf16>, memref<256x256xf16>) -> ()
    return
  }
```

</details>

Reply:
We are not talking about the same thing.
1. I am asking why use `tileConsumerAndFuseProducersUsingSCF` only and not use `tileAndFuseConsumerOfSlice` instead of `IterativeTilingAndFusion`.(this is the upstream method proposed by IREE guys and not the downstream one)
2. The branch I mentioned is not if-else branch but more like the residual branch in resnet. But indeed, your implementation performs better at the if-else branch case.

```
    C
  /  \
D      E
C = matmul(A, B)
D = ReLU(C)
E = LayerNorm(C)
```